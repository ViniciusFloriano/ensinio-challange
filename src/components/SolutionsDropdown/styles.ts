import styled from 'styled-components';

const Container = styled('div')`
  background-color: ${(props) => props.theme.colors.backgroundSecondary};
  border: 1px solid ${(props) => props.theme.colors.primary};

  display: flex;
  flex-direction: column;
  cursor: default;
  border-radius: 6px;
  padding: 1rem 0;

  h3 {
    padding: 1rem 0;
    padding-left: 4rem;
    text-transform: uppercase;
    text-align: left;
    width: 100%;
  }
  > div {
    text-align: left;
    width: 580px;
    display: flex;
    justify-content: space-between;
    padding: 0 2rem;
    flex-wrap: wrap;
    h1,
    p {
      width: 100%;
    }
  }
`;

export default Container;
