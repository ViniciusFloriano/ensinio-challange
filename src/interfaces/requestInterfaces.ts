import { LanguagesProps } from './languageInterfaces';

export interface ItemProps {
  id: number;
  title: LanguagesProps;
  description: LanguagesProps;
}
