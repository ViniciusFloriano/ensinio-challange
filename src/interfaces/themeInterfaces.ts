export interface ThemeProps {
  colors: {
    white?: string;
    black?: string;
    gray?: string;
    lightGray?: string;
  };
  sizes: {
    xsmall: string | number;
    small: string | number;
    normal: string | number;
    large: string | number;
    xlarge: string | number;
    xxlarge: string | number;
  };
}
