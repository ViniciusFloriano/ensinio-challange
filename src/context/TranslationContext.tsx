import React, { createContext, useEffect, useState } from 'react';

import { LanguagesProps } from '../interfaces/languageInterfaces';
import { TranslationProps } from '../interfaces/requestInterfaces';
import { getTranslations } from '../services/requests';

type LanguagesType = keyof LanguagesProps;

interface TranslationContextProps {
  loading: boolean;
  error: boolean;
  currentLang: LanguagesType;
  setCurrentLang: (language: LanguagesType) => void;
  t: (title: string) => string;
  translationData: TranslationProps[];
}

const Context = createContext<TranslationContextProps>({
  loading: true,
  error: false,
  currentLang: 'pt',
  setCurrentLang: () => {},
  t: (title: string) => title,
  translationData: [],
});

function useTranslation() {
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(true);
  const [translationData, setTranslationData] = useState<TranslationProps[]>(
    []
  );
  const [currentLang, setCurrentLangState] = useState<LanguagesType>('pt');

  useEffect(() => {
    getTranslations()
      .then((data) => {
        setTranslationData(data);
        setLoading(false);
      })
      .catch(() => {
        setError(true);
      });

    const languageStorage = localStorage.getItem('language');

    if (languageStorage) {
      setCurrentLangState(JSON.parse(languageStorage));
    } else {
      const navigatorLanguage = navigator.language;
      if (navigatorLanguage.includes('en')) {
        setCurrentLangState('en');
      } else if (navigatorLanguage.includes('pt')) {
        setCurrentLangState('pt');
      } else if (navigatorLanguage.includes('es')) {
        setCurrentLangState('es');
      }
    }
  }, []);

  function t(title: string) {
    const translatedText = translationData.find(
      (td) => td.title.toLowerCase() === title.toLowerCase()
    );

    if (!translatedText) {
      return 'Translation not find';
    }

    return translatedText.value[currentLang];
  }

  function setCurrentLang(language: LanguagesType) {
    localStorage.setItem('language', JSON.stringify(language));
    setCurrentLangState(language);
  }

  return { loading, error, currentLang, setCurrentLang, t, translationData };
}

const TranslationProvider: React.FC = ({ children }) => {
  const {
    loading,
    error,
    currentLang,
    setCurrentLang,
    t,
    translationData,
  } = useTranslation();

  return (
    <Context.Provider
      value={{
        loading,
        error,
        currentLang,
        setCurrentLang,
        t,
        translationData,
      }}
    >
      {children}
    </Context.Provider>
  );
};

export { Context as TranslationContext, TranslationProvider };
