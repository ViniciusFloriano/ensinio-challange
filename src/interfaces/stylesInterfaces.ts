import { ThemeProps } from './themeInterfaces';

export interface HeadingProps {
  color?: keyof ThemeProps['colors'];
  size?: keyof ThemeProps['sizes'];
  fontWeight?: 100 | 400 | 700;
  lineHeight?: string | number;
  level?: 1 | 2 | 3 | 4 | 5 | 6;
}
