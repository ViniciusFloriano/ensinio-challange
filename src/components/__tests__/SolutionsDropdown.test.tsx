import React from 'react';
import renderer from 'react-test-renderer';

import { cleanup } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';

import { TranslationContext } from '../../context/TranslationContext';
import { LanguagesProps } from '../../interfaces/languageInterfaces';
import GlobalStyle from '../../styles/global';
import defaultTheme from '../../styles/themes';
import SolutionsDropdown from '../SolutionsDropdown';

it('renders correctly when there are no items', () => {
  const language: keyof LanguagesProps = 'pt';
  const translation = {
    loading: true,
    error: false,
    currentLang: language,
    setCurrentLang: () => {},
    t: (title: string) => title,
    translationData: [],
  };
  const solutionsItems = [
    {
      key: '01',
      desciption: translation.t('Lorem ipsum dolor sit amet'),
      title: translation.t('Create an Online School'),
      icon: 'EadIcon',
    },
    {
      key: '02',
      desciption: translation.t('Lorem ipsum dolor sit amet'),
      title: translation.t('Community and social network'),
      icon: 'AppIcon',
    },
    {
      key: '03',
      desciption: translation.t('Lorem ipsum dolor sit amet'),
      title: translation.t('Gamification'),
      icon: 'GamificationIcon',
    },
    {
      key: '04',
      desciption: translation.t('Lorem ipsum dolor sit amet'),
      title: translation.t('Mobile app'),
      icon: 'SocialIcon',
    },
  ];

  const tree = renderer
    .create(
      <TranslationContext.Provider value={translation}>
        <ThemeProvider theme={defaultTheme}>
          <SolutionsDropdown
            items={solutionsItems}
            title="Solutions Dropdown"
          />
          <GlobalStyle />
        </ThemeProvider>
      </TranslationContext.Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
afterEach(cleanup);
