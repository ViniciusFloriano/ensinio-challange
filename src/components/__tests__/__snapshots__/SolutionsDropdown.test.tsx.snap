// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly when there are no items 1`] = `
<div
  className="sc-hKFxyN YvNCz"
>
  <h3
    className="sc-bdnxRM gLMxZa"
    color="linkPrimaryOpacity"
    fontFamily="fontSecondary"
    fontWeight={500}
    letterSpacing="0.12rem"
    size="xsmall"
  >
    Solutions Dropdown
  </h3>
  <div>
    <div
      className="sc-dlnjwi yojSu"
    >
      <img
        alt="Create an Online School"
        src="EadIcon"
      />
      <div>
        <h1
          className="sc-bdnxRM jfpdHn"
          color="textTertiary"
          size="xsmall"
        >
          Create an Online School
        </h1>
        <p
          className="sc-gtsrHT jwszLL"
          color="textTertiaryOpacity"
          fontWeight={400}
          size="xsmall"
        >
          Lorem ipsum dolor sit amet
        </p>
      </div>
    </div>
    <div
      className="sc-dlnjwi yojSu"
    >
      <img
        alt="Community and social network"
        src="AppIcon"
      />
      <div>
        <h1
          className="sc-bdnxRM jfpdHn"
          color="textTertiary"
          size="xsmall"
        >
          Community and social network
        </h1>
        <p
          className="sc-gtsrHT jwszLL"
          color="textTertiaryOpacity"
          fontWeight={400}
          size="xsmall"
        >
          Lorem ipsum dolor sit amet
        </p>
      </div>
    </div>
    <div
      className="sc-dlnjwi yojSu"
    >
      <img
        alt="Gamification"
        src="GamificationIcon"
      />
      <div>
        <h1
          className="sc-bdnxRM jfpdHn"
          color="textTertiary"
          size="xsmall"
        >
          Gamification
        </h1>
        <p
          className="sc-gtsrHT jwszLL"
          color="textTertiaryOpacity"
          fontWeight={400}
          size="xsmall"
        >
          Lorem ipsum dolor sit amet
        </p>
      </div>
    </div>
    <div
      className="sc-dlnjwi yojSu"
    >
      <img
        alt="Mobile app"
        src="SocialIcon"
      />
      <div>
        <h1
          className="sc-bdnxRM jfpdHn"
          color="textTertiary"
          size="xsmall"
        >
          Mobile app
        </h1>
        <p
          className="sc-gtsrHT jwszLL"
          color="textTertiaryOpacity"
          fontWeight={400}
          size="xsmall"
        >
          Lorem ipsum dolor sit amet
        </p>
      </div>
    </div>
  </div>
</div>
`;
